File: Leetcode_POTD_2025/7_JULY/1865_Finding_Pairs_With_a_Certain_Sum.cpp
class FindSumPairs {
    unordered_map<int,int> cache;
    vector<int> cpu, gpu;
public:
    FindSumPairs(vector<int>& a, vector<int>& b): cpu(a), gpu(b) {
        for(int i : b) cache[i]++;
    }
    void add(int idx, int val) {
        cache[gpu[idx]]--;
        gpu[idx] += val;
        cache[gpu[idx]]++;
    }
    int count(int tgt) {
        int cnt = 0;
        for(int x : cpu) cnt += cache[tgt - x];
        return cnt;
    }
};

File: Leetcode_POTD_2025/7_JULY/1865_Finding_Pairs_With_a_Certain_Sum.py
class FindSumPairs(object):
    def __init__(self, cpu, gpu):
        self.core = cpu
        self.chip = gpu
        self.freq = Counter(gpu)

    def add(self, port, inc):
        self.freq[self.chip[port]] -= 1
        self.chip[port] += inc
        self.freq[self.chip[port]] += 1

    def count(self, target):
        return sum(self.freq[target - val] for val in self.core)
